features/device_key/TESTS/device_key/functionality/main.cpp

```diff
@@ -0,0 +1,492 @@
+#define MSG_KEY_DEVICE_TEST_STEP3 "check_consistency_step3"
+#define MSG_KEY_DEVICE_TEST_STEP4 "check_consistency_step4"
+#define MSG_KEY_DEVICE_TEST_SUITE_ENDED "Test suite ended"
+
+void generate_derived_key_consistency_16_byte_key_reset_test(char *key);
+void generate_derived_key_consistency_32_byte_key_reset_test(char *key);
+
+/*
+ * Injection of a dummy key when there is no TRNG
+ */
+int inject_dummy_rot_key()
+{
+#if !defined(DEVICE_TRNG)
+    uint32_t key[DEVICE_KEY_16BYTE / sizeof(uint32_t)];
+
+    memset(key, 0, DEVICE_KEY_16BYTE);
+    memcpy(key, "1234567812345678", DEVICE_KEY_16BYTE);
+    int size = DEVICE_KEY_16BYTE;
+    DeviceKey& devkey = DeviceKey::get_instance();
+    return devkey.device_inject_root_of_trust(key, size);
+#else
+    return DEVICEKEY_SUCCESS;
+#endif
+}
+
+void generate_derived_key_reset_test()
+{
+    greentea_send_kv(MSG_KEY_DEVICE_READY, MSG_VALUE_DUMMY);
+
+    static char key[MSG_KEY_LEN + 1] = { };
+    static char value[MSG_VALUE_LEN + 1] = { };
+    memset(key, 0, MSG_KEY_LEN + 1);
+    memset(value, 0, MSG_VALUE_LEN + 1);
+
+    greentea_parse_kv(key, value, MSG_KEY_LEN, MSG_VALUE_LEN);
+
+    if (strcmp(key, MSG_KEY_DEVICE_TEST_STEP1) == 0 || strcmp(key, MSG_KEY_DEVICE_TEST_STEP2) == 0) {
+        generate_derived_key_consistency_16_byte_key_reset_test(key);
+        return generate_derived_key_reset_test();
+    }
+
+    if (strcmp(key, MSG_KEY_DEVICE_TEST_STEP3) == 0 || strcmp(key, MSG_KEY_DEVICE_TEST_STEP4) == 0) {
+        return generate_derived_key_consistency_32_byte_key_reset_test(key);
+    }
+
+    TEST_ASSERT_MESSAGE(false, key); //Indicates error!!!
+}
+
+/*
+ * Test the consistency of derived 16 byte key result after device reset.
+ */
+void generate_derived_key_consistency_16_byte_key_reset_test(char *key)
+{
+    unsigned char output1[DEVICE_KEY_16BYTE];
+    unsigned char output2[DEVICE_KEY_16BYTE];
+    unsigned char empty_buffer[DEVICE_KEY_16BYTE];
+    unsigned char salt[] = "Once upon a time, I worked for the circus and I lived in Omaha.";
+    int key_type = DEVICE_KEY_16BYTE;
+    uint16_t actual_size = 0;
+    DeviceKey& devkey = DeviceKey::get_instance();
+    NVStore& nvstore = NVStore::get_instance();
+    size_t salt_size = sizeof(salt);
+
+    if (strcmp(key, MSG_KEY_DEVICE_TEST_STEP1) == 0) {
+
+        //Third step: Clear NVStore, create an ROT key, derive a 16 byte
+        //key and store it in NVStore at index 15, At the end reset the device
+        int ret = nvstore.reset();
+        TEST_ASSERT_EQUAL_INT(DEVICEKEY_SUCCESS, ret);
+
+        ret = inject_dummy_rot_key();
+        TEST_ASSERT_EQUAL_INT(DEVICEKEY_SUCCESS, ret);
+
+        memset(output1, 0, sizeof(output1));
+        ret = devkey.generate_derived_key(salt, salt_size, output1, key_type);
+        TEST_ASSERT_EQUAL_INT32(0, ret);
+        bool is_empty = !memcmp(empty_buffer, output1, sizeof(output1));
+        TEST_ASSERT_FALSE(is_empty);
+
+        ret = nvstore.set(15, DEVICE_KEY_16BYTE, output1);
```

Comment:
@yossi2le What's the relevant of the constant `15` here?

Reply:
@hanno-arm 
15 is the key for the last entry in nvstore. During the test we use this entry to store the derived key and then we read it after reset. We expect it to be equal to a new key derived from the same salt.

Reply:
@yossi2le Thanks!